database/db.go

```diff
@@ -0,0 +1,284 @@
+	// Check if a skylink with this hash already exists in the DB.
+	sr := db.Skylinks.FindOne(ctx, bson.M{"hash": skylink.Hash})
+	if sr.Err() != mongo.ErrNoDocuments {
+		return ErrSkylinkExists
+	}
+	_, err := db.Skylinks.InsertOne(ctx, skylink)
+	if err != nil && strings.Contains(err.Error(), "E11000 duplicate key error collection: scanner.skylinks index: skylink_unique") {
+		// This skylink already exists in the DB.
+		return nil
+	}
+	return err
+}
+
+// SkylinkSave saves the given Skylink record to the database.
+func (db *DB) SkylinkSave(ctx context.Context, skylink *Skylink) error {
+	filter := bson.M{"_id": skylink.ID}
+	opts := &options.ReplaceOptions{
+		Upsert: &True,
+	}
+	_, err := db.Skylinks.ReplaceOne(ctx, filter, skylink, opts)
+	if err != nil {
+		return errors.AddContext(err, "failed to save")
+	}
+	return nil
+}
+
+// CancelStuckScans resets the status of scans that have been going on for more
+// than scanner.CancelScanAfter. We assume that these scans have terminated
+// unexpectedly without reporting their results (e.g. server crash).
+func (db *DB) CancelStuckScans(ctx context.Context) (int64, error) {
+	filter := bson.M{
+		"status":    SkylinkStatusScanning,
+		"timestamp": bson.M{"$gt": time.Now().Add(-CancelScanAfter)},
+	}
+	update := bson.M{
+		"$set": bson.M{
+			"timestamp": time.Now().UTC(),
+			"status":    SkylinkStatusNew,
+		},
+	}
+	ur, err := db.Skylinks.UpdateMany(ctx, filter, update)
+	if err != nil {
+		return 0, err
+	}
+	return ur.ModifiedCount, nil
+}
+
+// SweepAndLock sweeps the database for new skylinks. It "locks" and returns the
+// first one it encounters. The "locking" is done by updating the skylink's
+// status from "new" to "scanning".
+func (db *DB) SweepAndLock(ctx context.Context) (*Skylink, error) {
+	filter := bson.M{"status": SkylinkStatusNew}
+	update := bson.M{
+		"$set": bson.M{
+			"timestamp": time.Now().UTC(),
+			"status":    SkylinkStatusScanning,
+		},
+	}
+	// Look for a single new record and change its status to "scanning".
+	// ur, err := db.Skylinks.UpdateOne(ctx, filter, update)
+	sr := db.Skylinks.FindOneAndUpdate(ctx, filter, update)
+	if sr.Err() == mongo.ErrNoDocuments {
+		return nil, ErrNoDocumentsFound
+	}
+	if sr.Err() != nil {
+		return nil, sr.Err()
+	}
+	var sl Skylink
+	err := sr.Decode(&sl)
+	if err != nil {
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// ensureDBSchema checks that we have all collections and indexes we need and
+// creates them if needed.
+// See https://docs.mongodb.com/manual/indexes/
+// See https://docs.mongodb.com/manual/core/index-unique/
+func ensureDBSchema(ctx context.Context, db *mongo.Database, log *logrus.Logger) error {
+	// schema defines a mapping between a collection name and the indexes that
+	// must exist for that collection.
+	schema := map[string][]mongo.IndexModel{
+		dbSkylinks: {
+			{
+				Keys:    bson.D{{"skylink", 1}},
+				Options: options.Index().SetName("skylink"),
```

Comment:
No. We don't want to keep the skylinks in the DB because they might be pointing to illegal content. All we want to keep is their hashes. This means that after scanning a skylink, we'll replace it with an empty string (or delete it from the record) and that will clash with this constraint.